core/services/pipeline/scheduler.go

```diff
@@ -72,7 +153,7 @@ Loop:
 						Task:       task,
 						Result:     Result{Error: ErrTimeout},
 						CreatedAt:  now, // TODO: more accurate start time
-						FinishedAt: now,
+						FinishedAt: &now,
```

Comment:
This function is pretty beefy, can it be split out into two or three smaller chunks?